// gdxts/dist/lib/Glyph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Glyph = exports.PAGES = exports.PAGE_SIZE = exports.LOG2_PAGE_SIZE = void 0;
exports.LOG2_PAGE_SIZE = 9;
exports.PAGE_SIZE = 1 << exports.LOG2_PAGE_SIZE;
exports.PAGES = 0x10000 / exports.PAGE_SIZE;
class Glyph {
    constructor() {
        /** The index to the texture page that holds this glyph. */
        this.page = 0;
    }
    getKerning(ch) {
        if (this.kerning) {
            const page = this.kerning[ch.charCodeAt(0) >>> exports.LOG2_PAGE_SIZE];
            if (page != null)
                return page[ch.charCodeAt(0) & (exports.PAGE_SIZE - 1)];
        }
        return 0;
    }
    setKerning(ch, value) {
        if (!this.kerning) {
            this.kerning = [];
        }
        let page = this.kerning[ch >>> exports.LOG2_PAGE_SIZE];
        if (!page) {
            this.kerning[ch >>> exports.LOG2_PAGE_SIZE] = page = new Uint32Array(new ArrayBuffer(exports.PAGE_SIZE));
        }
        page[ch & (exports.PAGE_SIZE - 1)] = value;
    }
}
exports.Glyph = Glyph;
